features/323-login-id/design.md

```diff
@@ -0,0 +1,109 @@
+# Background
+
+## Current Design
+
+Under current design of authentication:
+
+1. Each **User** is associated with some **Principal**.
+3. Each **Principal** is associated with one **Identity Provider** (e.g. password authentication, OAuth).
+4. Password authentication principals is represented by tuple (**Login ID Key**, **Login ID**, password hash).
+5. App developer specifies a list of white-listed login ID key.
+6. To signup using password authentication, user can provide a list of tuple (login ID key, login ID). e.g. `{ "username": "test", "email": "a@example.com" }`
+7. For password authentication principals, user can login using a tuple (login ID key, login ID). e.g. `{ "email": "a@example.com" }`
+8. Login IDs are globally unique in app, without considering associated login ID keys.
+9. Auth gear does not knows the semantics of login IDs.
+
+
+## Problem
+1. Some features requires semantics of login IDs to function.
+2. Some authentication system designs may not be implemented easily in current design.
+
+
+# Use cases
+
+Considering some practical scenarios of authentication system:
+
+## GitHub
+- Supports multiple emails per user, with email verification.
+- User can add/delete any amount of emails.
+- Only verified emails would be associated with the user in public.
+- User can allow recovery email/security alert to be sent to all verified emails or a specific verified email.
+
+
+## Facebook
+- User can login with email/phone number/username, without indicating what type of value it is.
+- User can add/delete any amount of emails, which can be used to login after verification.
+- Disposable email address is blocked.
+- All emails/phones of user is verified.
+- Email communications are personalized with user language & name
+
+
+# Design Objectives
+- Allow auth related features to understand semantics of some login IDs / user metadata
+- Support multiple login ID with same login ID key
+- Support adding/removing login ID
```

Comment:
[As specified in #294](https://github.com/SkygearIO/features/blob/master/features/294-login-without-specifying-login-id-key/api.md), there are `createLoginID`, `updateLoginID` and `removeLoginID`, so this objective should be in the existing design already.